File: src/main/webapp/js/estudiante.js
/**
 * Clase Estudiante (Programación Orientada a Objetos en JavaScript).
 * Encapsula los datos y la lógica de cálculo de un estudiante, para que
 * el resto del código (app.js) solo tenga que crear objetos y leer sus
 * resultados, sin repetir las reglas de negocio.
 */
class Estudiante {

    static NOTA_MINIMA = 0.0;
    static NOTA_MAXIMA = 5.0;

    constructor(nombre, nota1, nota2, nota3, nota4, id = null) {
        this.id = id;
        this.nombre = nombre;
        this.nota1 = nota1;
        this.nota2 = nota2;
        this.nota3 = nota3;
        this.nota4 = nota4;
        this.promedio = 0.0;
        this.resultado = "";
    }

    /** Calcula el promedio de las cuatro notas, redondeado a un decimal y limitado a [0.0, 5.0]. */
    calcularPromedio() {
        const suma = this.nota1 + this.nota2 + this.nota3 + this.nota4;
        let valor = suma / 4;
        valor = Math.round(valor * 10) / 10;
        this.promedio = Math.max(Estudiante.NOTA_MINIMA, Math.min(Estudiante.NOTA_MAXIMA, valor));
        return this.promedio;
    }

    /** El estudiante aprueba si el promedio es igual o superior a 3.0. */
    determinarAprobacion() {
        return this.promedio >= 3.0;
    }

    /** Traduce el promedio cuantitativo a la valoración cualitativa. */
    determinarRendimiento() {
        if (this.promedio < 3.0) {
            this.resultado = "Rendimiento insuficiente (R.I)";
        } else if (this.promedio < 4.0) {
            this.resultado = "Aprobado (A)";
        } else if (this.promedio <= 4.5) {
            this.resultado = "Aprobado con sobresaliente (A.S)";
        } else {
            this.resultado = "Aprobado con excelente (A.E)";
        }
        return this.resultado;
    }

    /** Devuelve la clase CSS (badge de color) correspondiente al resultado. */
    claseResultado() {
        if (this.resultado.includes("R.I")) return "ri";
        if (this.resultado.includes("A.S")) return "as";
        if (this.resultado.includes("A.E")) return "ae";
        return "a";
    }

    /**
     * Valida una nota individual: obligatoria, numérica, sin negativos y
     * sin superar 5.0. Devuelve un mensaje de error o null si es válida.
     */
    static validarNota(valorTexto, etiqueta) {
        if (valorTexto === "" || valorTexto === null || valorTexto === undefined) {
            return `${etiqueta} es obligatoria.`;
        }
        // Un solo punto decimal, sin signo negativo.
        if (!/^\d+(\.\d+)?$/.test(String(valorTexto).trim())) {
            return `${etiqueta} debe ser un número positivo válido.`;
        }
        const nota = parseFloat(valorTexto);
        if (Number.isNaN(nota)) {
            return `${etiqueta} debe ser un valor numérico.`;
        }
        if (nota < Estudiante.NOTA_MINIMA) {
            return `${etiqueta} no puede ser negativa.`;
        }
        if (nota > Estudiante.NOTA_MAXIMA) {
            return `${etiqueta} no puede ser mayor que 5.0.`;
        }
        return null;
    }
}

File: src/main/java/com/notas/model/Estudiante.java
package com.notas.model;

/**
 * Modelo que representa a un estudiante y sus calificaciones.
 * Contiene la logica de calculo (promedio, aprobacion y rendimiento
 * cualitativo) para que el Servlet y el DAO trabajen siempre con un
 * objeto ya validado y calculado, sin duplicar reglas de negocio.
 */
public class Estudiante {

    public static final double NOTA_MINIMA = 0.0;
    public static final double NOTA_MAXIMA = 5.0;

    private int id;
    private String nombre;
    private double nota1;
    private double nota2;
    private double nota3;
    private double nota4;
    private double promedio;
    private String resultado;

    public Estudiante() {
    }

    public Estudiante(String nombre, double nota1, double nota2, double nota3, double nota4) {
        this.nombre = nombre;
        this.nota1 = nota1;
        this.nota2 = nota2;
        this.nota3 = nota3;
        this.nota4 = nota4;
        calcularPromedio();
        determinarRendimiento();
    }

    /** Calcula el promedio de las cuatro notas y lo limita al rango 0.0 - 5.0. */
    public double calcularPromedio() {
        double suma = nota1 + nota2 + nota3 + nota4;
        double valor = suma / 4.0;
        valor = Math.round(valor * 10.0) / 10.0;
        this.promedio = Math.max(NOTA_MINIMA, Math.min(NOTA_MAXIMA, valor));
        return this.promedio;
    }

    /** Un estudiante aprueba cuando el promedio es igual o superior a 3.0. */
    public boolean determinarAprobacion() {
        return this.promedio >= 3.0;
    }

    /** Traduce el promedio cuantitativo a la valoracion cualitativa de la guia. */
    public String determinarRendimiento() {
        if (promedio < 3.0) {
            resultado = "Rendimiento insuficiente (R.I)";
        } else if (promedio < 4.0) {
            resultado = "Aprobado (A)";
        } else if (promedio <= 4.5) {
            resultado = "Aprobado con sobresaliente (A.S)";
        } else {
            resultado = "Aprobado con excelente (A.E)";
        }
        return resultado;
    }

    /**
     * Valida que las cuatro notas sean numeros dentro del rango permitido
     * (sin negativos y sin superar 5.0). Devuelve un mensaje de error o
     * null si todo es valido.
     */
    public static String validarNota(double nota, String etiqueta) {
        if (Double.isNaN(nota)) {
            return etiqueta + " debe ser un valor numerico.";
        }
        if (nota < NOTA_MINIMA) {
            return etiqueta + " no puede ser negativa.";
        }
        if (nota > NOTA_MAXIMA) {
            return etiqueta + " no puede ser mayor que 5.0.";
        }
        return null;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public double getNota1() {
        return nota1;
    }

    public void setNota1(double nota1) {
        this.nota1 = nota1;
    }

    public double getNota2() {
        return nota2;
    }

    public void setNota2(double nota2) {
        this.nota2 = nota2;
    }

    public double getNota3() {
        return nota3;
    }

    public void setNota3(double nota3) {
        this.nota3 = nota3;
    }

    public double getNota4() {
        return nota4;
    }

    public void setNota4(double nota4) {
        this.nota4 = nota4;
    }

    public double getPromedio() {
        return promedio;
    }

    public String getResultado() {
        return resultado;
    }
}
